.github/workflows/clients.yml

```diff
@@ -0,0 +1,70 @@
+name: Clients
+
+concurrency:
+  group: clients-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+on:
+  workflow_call:
+
+jobs:
+  client-dotnet:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest, [self-hosted, ARM64, macos-12.6], [self-hosted, ARM64, macos-13.2]]
+
+      runs-on: ${{ matrix.os }}
+      steps:
+        - uses: actions/checkout@v3
+        - uses: actions/setup-dotnet@v3
+        - run: ./scripts/install_zig.sh
+        - run: ./zig/zig build client_ci -- --language=dotnet
+
+  client-go:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest, [self-hosted, ARM64, macos-12.6], [self-hosted, ARM64, macos-13.2]]
+
+      runs-on: ${{ matrix.os }}
+      steps:
+        - uses: actions/checkout@v3
+        - uses: actions/setup-go@v4
+        - run: ./scripts/install_zig.sh
+        - run: ./zig/zig build client_ci -- --language=go
+
+  client-java:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest, [self-hosted, ARM64, macos-12.6], [self-hosted, ARM64, macos-13.2]]
+
+      runs-on: ${{ matrix.os }}
+      steps:
+        - uses: actions/checkout@v3
+        - uses: actions/setup-java@v3
+        - run: ./scripts/install_zig.sh
+        - run: ./zig/zig build client_ci -- --language=java
+
+  client-node:
+    strategy:
+      matrix:
+        os: [ubuntu-latest, windows-latest, macos-latest, [self-hosted, ARM64, macos-12.6], [self-hosted, ARM64, macos-13.2]]
+
+      runs-on: ${{ matrix.os }}
+      steps:
+        - uses: actions/checkout@v3
+        - uses: actions/setup-node@v3
+        - run: ./scripts/install_zig.sh
+        - run: ./zig/zig build client_ci -- --language=node
```

Comment:
I guess this is the core thing here, conceptually. This is how I relatively strongly feel our CI should look: define the matrix in `.yaml`, use upstream dependencies (`setup-x` actions), just like how our users would, launch the zig process and then stack away from yml.